<launch>

  <!-- <arg name="odom_topic" default="odom_noisy"/> -->
  <arg name="odom_topic" default="odom_noisy"/>
  <arg name="got_topic" default="got_pose"/>
  <arg name="x_init" default="2.0"/>
  <arg name="y_init" default="-4.0"/>
  <arg name="z_init" default="0.0"/>
  <!-- <arg name="yaw_init" default="3.14"/> -->
  <arg name="yaw_init" default="0.0"/>
  <arg name="odometry_filtered_topic" default="odometry/filtered_odom"/>



  <!-- Launch got_node publishing geometry_msgs/PoseWithCovarianceStamped -->
  <!-- <node pkg="got_node" type="got_position" name="got_position_node" /> -->


<!-- So, to estimate and publish both the map → odom and the odom → base_linktransforms (or state estimates), we need two robot_localization state estimators:

    one for the odom → base_link transform, which gets all continuous inputs (acceleration and linear and angular velocity) and
    one for the map → odom transform, which gets all (both continuous and non-continuous) inputs (acceleration, linear and angular velocity, position and orientation). -->


  <!-- odom -> base_link -->
  <!-- robot_localization EKF node for the odom frame -->
  <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_odom" clear_params="true">
    <param name="frequency" value="10." />  
    <param name="sensor_timeout" value="0.2" />  
    <param name="two_d_mode" value="true" />
    <param name="publish_tf" value="true" />
    <param name="map_frame" value="map" />
    <param name="odom_frame" value="odom" />
    <param name="base_link_frame" value="base_footprint" />
    <param name="world_frame" value="odom" />
    <param name="print_diagnostics" value="true" />
    <remap from="odometry/filtered" to="$(arg odometry_filtered_topic)" />
    <!-- <rosparam param="initial_state">[$(arg x_init),  $(arg y_init),  $(arg z_init), -->
    <!-- <rosparam param="initial_state">[2.0,  -4.0,  0.0, -->
    <rosparam param="initial_state">[0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0]</rosparam>
    <param name="odom0" value="$(arg odom_topic)" />
    <param name="odom0_differential" value="true"/>
    <rosparam param="odom0_config">[false, false, false,
                                    false, false, false,
                                    true, true, false,
                                    false, false, true,
                                    false, false, false]</rosparam>
<!--     <param name="imu0" value="imu" />
    <param name="imu0_differential" value="false"/>
    <rosparam param="imu0_config">[false, false, false,
                                   false, false, true,
                                   false, false, false,
                                   false, false, true,
                                   true, false, false]</rosparam> -->
<!--     <param name="imu0" value="imu" />
    <param name="imu0_differential" value="false"/>
    <rosparam param="imu0_config">[false, false, false,
                                   false, false, true,
                                   false, false, false,
                                   false, false, false,
                                   false, false, false]</rosparam> -->
    <param name="pose0" value="$(arg got_topic)" />
    <param name="pose0_differential" value="true"/>
    <rosparam param="pose0_config">[true, true, false,
                                  false, false, false,
                                  false, false, false,
                                  false, false, false,
                                  false, false, false]</rosparam>

<!--     <rosparam param="process_noise_covariance">[0.01, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0.01, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0.01, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0.03, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0.03, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0.06, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0.05, 0,     0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0.05, 0,    0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0.06, 0,    0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.05, 0,    0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.05, 0,    0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0,
                                                0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam> -->
  </node>

<!--   <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_map" clear_params="true">
    <param name="frequency" value="10." />  
    <param name="sensor_timeout" value="0.2" />  
    <param name="two_d_mode" value="true" />
    <param name="publish_tf" value="false" />
    <param name="map_frame" value="map" />
    <param name="odom_frame" value="odom" />
    <param name="base_link_frame" value="base_footprint" />
    <param name="world_frame" value="map" />
    <param name="print_diagnostics" value="true" />
    <remap from="odometry/filtered" to="odometry/filtered_map" />
    <rosparam param="initial_state">[2.0,  -4.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0]</rosparam>
    <param name="odom0" value="$(arg got_topic)" />
    <param name="odom0_differential" value="false"/>
  <rosparam param="odom0_config">[false, false, false,
                                  false, false, false,
                                  true, true, false,
                                  false, false, true,
                                  false, false, false]</rosparam>
    <param name="got0" value="$(arg got_topic)" />
    <param name="got0_differential" value="true"/>
    <rosparam param="got0_config">[true, true, false,
                                  false, false, false,
                                  false, false, false,
                                  false, false, false,
                                  false, false, false]</rosparam>
    <param name="imu0" value="imu" />
    <param name="imu0_differential" value="false"/>
    <rosparam param="imu0_config">[false, false, false,
                                   false, false, true,
                                   false, false, false,
                                   false, false, true,
                                   true, false, false]</rosparam>
  </node> -->

<!--   <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_map" clear_params="true">
    <param name="frequency" value="10." />  
    <param name="sensor_timeout" value="0.2" />  
    <param name="two_d_mode" value="true" />
    <param name="publish_tf" value="true" />
    <param name="map_frame" value="map" />
    <param name="odom_frame" value="odom" />
    <param name="base_link_frame" value="base_footprint" />
    <param name="world_frame" value="map" />
    <param name="print_diagnostics" value="true" />
    <remap from="odometry/filtered" to="odometry/filtered_map" />
    <param name="odom0" value="$(arg odom_topic)" />
    <param name="odom0_differential" value="false"/>
    <rosparam param="initial_state">[2.0,  -4.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0,
                                     0.0,  0.0,  0.0]</rosparam>
    <rosparam param="odom0_config">[true, true, false,
                                  false, false, true,
                                  true, false, false,
                                  false, false, true,
                                  false, false, false]</rosparam>
    <param name="imu0" value="imu" />
    <param name="imu0_differential" value="false"/>
    <rosparam param="imu0_config">[false, false, false,
                                   false, false, true,
                                   false, false, false,
                                   false, false, true,
                                   true, false, false]</rosparam>
  </node> -->

  <!-- map -> odom -->

</launch> 